File: RunSeason2/backend/424.fortune_number/optimize.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// Функция для суммирования цифр в первой и второй половинах числа
func sumOfHalfDigits(number string) (int, int) {
	half := len(number) / 2
	firstHalf := 0
	secondHalf := 0

	for i := 0; i < half; i++ {
		digit, _ := strconv.Atoi(string(number[i]))
		firstHalf += digit
	}

	for i := half; i < len(number); i++ {
		digit, _ := strconv.Atoi(string(number[i]))
		secondHalf += digit
	}

	return firstHalf, secondHalf
}

// Функция для проверки, равны ли суммы первой и второй половин
func isSumEqual(number string) bool {
	firstHalf, secondHalf := sumOfHalfDigits(number)
	return firstHalf == secondHalf
}

// Основная функция для нахождения ближайшего числа
func findClosestNumber(inputNumber string) string {
	number := strconv.Itoa(toInt(inputNumber) + 1)
	number = padLeft(number, len(inputNumber))

	for {
		if number != strings.Repeat("0", len(inputNumber)) && isSumEqual(number) {
			return number
		}
		number = strconv.Itoa(toInt(number) + 1)
		number = padLeft(number, len(inputNumber))

		if len(number) > len(inputNumber) {
			number = strings.Repeat("0", len(inputNumber))
		}
	}
}

// Функция для преобразования строки в целое число
func toInt(s string) int {
	result, _ := strconv.Atoi(s)
	return result
}

// Функция для дополнения строки нулями слева
func padLeft(s string, length int) string {
	return fmt.Sprintf("%0*s", length, s)
}

// Главная функция
func main() {
	var inputNumber string
	fmt.Scan(&inputNumber)
	fmt.Println(findClosestNumber(inputNumber))
}

File: RunSeasons/RunSeason2/backend/424.fortune_number/optimize.cpp
#include <iostream>
#include <string>
#include <cctype>

std::string sum_of_half_digits(const std::string& number) {
    int half = number.length() / 2;
    int first_half = 0, second_half = 0;
    for (int i = 0; i < half; i++) {
        first_half += number[i] - '0';
    }
    for (int i = half; i < number.length(); i++) {
        second_half += number[i] - '0';
    }
    return std::to_string(first_half) + " " + std::to_string(second_half);
}

bool is_sum_equal(const std::string& number) {
    std::string halves = sum_of_half_digits(number);
    int space = halves.find(" ");
    int first_half = std::stoi(halves.substr(0, space));
    int second_half = std::stoi(halves.substr(space + 1));
    return first_half == second_half;
}

std::string find_closest_number(const std::string& input_number) {
    std::string number = std::to_string(std::stoi(input_number) + 1);
    number = std::string(input_number.length() - number.length(), '0') + number;
    while (true) {
        if (number != std::string(input_number.length(), '0') && is_sum_equal(number)) {
            return number;
        }
        number = std::to_string(std::stoi(number) + 1);
        if (number.length() > input_number.length()) {
            number = std::string(input_number.length(), '0');
        }
    }
}

int main() {
    std::string number;
    std::getline(std::cin, number);
    std::cout << find_closest_number(number) << std::endl;
    return 0;
}
